scheduler/src/main/java/io/camunda/zeebe/scheduler/ActorControl.java

```diff
@@ -136,7 +136,7 @@ public ActorFuture<Void> call(final Runnable action) {
    * @return
    */
   public ScheduledTimer runDelayed(final Duration delay, final Runnable runnable) {
-    ensureCalledFromWithinActor("runDelayed(...)");
+//    ensureCalledFromWithinActor("runDelayed(...)");
```

Comment:
The timer will be added to the timer queue which is part of the ActorThread. Tbh I see not really an issue here, even if this is another thread. When the timer is due the ActorTask will be enqueued and run from some ActorThread :shrug: Maybe I oversee something. Do you think we can remove the restriction?

@npepinpe or @saig0 maybe you have an idea?

Reply:
IIRC the DeadlineTimerWheel (which the ActorTimerQueue extends) is not thread safe, so calling it from a different actor may cause race conditions.

Reply:
Ah good point so it is about the DeadlineTimerWheel, why we have to use our thread. So one possibility would be to make it thread safe? :thinking: 

Here btw is the check https://github.com/camunda/zeebe/blob/main/scheduler/src/main/java/io/camunda/zeebe/scheduler/ActorControl.java#L363-L377 where we verify whether it is our OWN actor :see_no_evil:

Reply:
Yeah, that's the one :smile: To me it's more confusing - I mean I know what it does, but I don't think it's obvious the first time you read it what it guarantees :smile:

Reply:
I don't know about making the actor timer queue thread safe. I guess there's no real harm other than maybe performance, but I wasn't there when this was decided and it's not documented, so maybe we'll find out there was a good reason :smile:

Reply:
@npepinpe another idea. What do you think of we keep the additional submit, to keep it thread-safe, but create an own actor for the _PRocessingScheduleService_. This might reduce the potential delays (since the actor only has to schedule and execute timers). This would also solve other issues, like blocking processing actor when having many timers or timeouts. If you agree then I would create another PR where I migrate the ProcessingScheduleService to an own actor with own

Reply:
Hm, so upside is that it's not blocking the processing actor anymore (which is nice), downside is that there is some communication between both actors (in the form of the abort condition, the batch writer, etc.) which probably gets more complicated. I don't think it's too crazy for it to have its own writer tbh, and maybe even own abort condition (which is controlled when closing the whole stream processor), so maybe it's not too bad. OTOH I think it's a good idea